overlord/hookstate/ctlcmd/helpers.go

```diff
@@ -149,7 +149,7 @@ func runServiceCommand(context *hookstate.Context, inst *servicestate.Instructio
 		return err
 	}
 
-	if !context.IsEphemeral() && context.HookName() == "configure" {
+	if !context.IsEphemeral() && (context.HookName() == "configure" || context.HookName() == "default-configure") {
```

Comment:
where do the tasks queued after default-configure exactly end up? it's not very transparent from tests

Reply:
They queued at the end after all install tasks
https://github.com/snapcore/snapd/blob/0bf3e39e00948e76ed45cbcc96facee8da181395/overlord/hookstate/ctlcmd/helpers.go#L114-L124

Reply:
I added clarifying comments in the test

Reply:
does this mean after the configure hook? does this solve the original issue people were having in ways they expect?

Reply:
Yes, after both the configure hook and the health-check hook.

For illustration this is the expected order of tasks for installing a snap which is used in the unit tests:
```
"prerequisites",
"download-snap",
"validate-snap",
"mount-snap",
"copy-snap-data",
"setup-profiles",
"link-snap",
"auto-connect",
"set-auto-aliases",
"setup-aliases",
"run-hook[install]",
"run-hook[default-configure]",
"start-snap-services",
"run-hook[configure]",
"run-hook[check-health]",
"exec-command",                      <----  task for snapctl command
"service-control",                       <----  task for service operation
```

Reply:
I see, we probably want to ask the reporter on the bugs if this is too late for them or ok?

Reply:
> does this solve the original issue people were having in ways they expect?

Yes, I tested it locally using the snap they provided. I left a comment there but I am waiting for feedback.

Reply:
> I see, we probably want to ask the reporter on the bugs if this is too late for them or ok?

Good point, I will check.

Reply:
Thanks @ZeyadYasser for the fix and your explanations.

The use case is from a snap that has a one-shot service doing several firewall and network configuration to set up a Thread bridge (linked in the bug report). As mentioned in https://github.com/snapcore/snapd/pull/13960#issuecomment-2133693371, the services are disabled by default, so that the user (on classic but also a pre-built Core image) gets a chance to configure for e.g. the networking interface before starting the services. However, the same snap needs to start automatically on Core when the config is seeded from a gadget. (A workaround was suggested by @zyga to keep services enabled by default but not run them when the needed config isn't set; this could work well for a limited config set.)

In our case, with the workflow proposed in this PR, it could work as follows:
1. `default-configure` hook calls `snapctl start` if the `autostart` config option is set to true.
2. `configure` hook does some validation (not done in `default-configure` to avoid redundant checks).
3. services start.

But I suppose this isn't valid in every use case, as some applications expect the `configure` hook to execute while the services are running.

In summary, while the proposed order isn't an issue for our use case, I believe starting the services right after the hook succeeds makes this more inline with the snap best practices.

Side note: I wonder why `run-hook[check-health]` is ordered between `configure` and `exec-command`/`service-control` tasks.

Another solution I can think of is to add the ability to just enable/disable services via `snapctl` without starting or stopping them. Then, the `configure-hook` could simply call for e.g. the `snapctl enable --service <service>` (`snapctl enable` does something else) command (without needing to queue) and the following task (`start-snap-services`) should just take care of the startup. I've already discussed this with @ernestl, and I'm not sure if this is going to work. I have experimented by calling `snapctl stop --disable <service>` from `install` and `default-configure` hooks for an **enabled** service, and it works as expected: the service gets disabled and never starts.

Reply:
the earliest we can start those services is after snap-start-services,  @ZeyadYasser could you look what would be the complexity of queuing commands from default-configure there?

Reply:
We could have a variant of `queueCommand` or some special handling code that does the following:
- if the service commands are coming from `default-configure` hook
- simply inject the commands' tasksets after `start-snap-services` using `snapstate.InjectTasks`
    - `snapstate.InjectTasks` takes care of the complexity of adjusting the tasks' dependencies

Reply:
The PR is now updated to inject service commands after `start-snap-services` for the `default-configure` hook.